File: specs/typespec/org/agency-referrals.ts
import type { Handle } from "../hub/hub-users.js";

export type ApplicationMode = "open" | "agency_only";

export type AgencyReferralState =
	| "pending"
	| "accepted_applied"
	| "declined"
	| "expired"
	| "not_selected";

// ---- Consumer side: assign / list / remove agencies on an opening ----

export interface AssignOpeningAgencyRequest {
	opening_id: string;
	agency_org_domain: string;
}

export interface RemoveOpeningAgencyRequest {
	opening_id: string;
	agency_org_domain: string;
}

export interface ListOpeningAgenciesRequest {
	opening_id: string;
}

export interface OpeningAgency {
	agency_org_domain: string;
	agency_org_name: string;
	assigned_at: string;
	referrals_made: number;
}

export interface ListOpeningAgenciesResponse {
	agencies: OpeningAgency[];
}

// ---- Agency side: openings I'm assigned to ----

export interface ListAssignedOpeningsRequest {
	pagination_key?: string;
	limit?: number;
}

export interface AssignedOpening {
	opening_id: string;
	consumer_org_domain: string;
	opening_number: number;
	title: string;
	assigned_at: string;
}

export interface ListAssignedOpeningsResponse {
	openings: AssignedOpening[];
	next_pagination_key?: string;
}

// ---- Agency side: refer a candidate ----

export interface ReferCandidateRequest {
	opening_id: string;
	candidate_handle: Handle;
	statement_text?: string;
}

export interface ReferCandidateResponse {
	referral_id: string;
}

// ---- Agency side: referrals my agency has made ----

export interface ListAgencyReferralsRequest {
	pagination_key?: string;
	limit?: number;
}

export interface AgencyReferral {
	referral_id: string;
	candidate_handle: Handle;
	consumer_org_domain: string;
	opening_number: number;
	opening_title: string;
	state: AgencyReferralState;
	created_at: string;
}

export interface ListAgencyReferralsResponse {
	referrals: AgencyReferral[];
	next_pagination_key?: string;
}

export interface ValidationError {
	field: string;
	message: string;
}

function reqObj(req: unknown): Record<string, unknown> | null {
	if (!req || typeof req !== "object") return null;
	return req as Record<string, unknown>;
}

export function validateAssignOpeningAgencyRequest(
	req: unknown
): ValidationError[] {
	const r = reqObj(req);
	if (!r) return [{ field: "$root", message: "Request body is required" }];
	const errors: ValidationError[] = [];
	if (typeof r.opening_id !== "string" || r.opening_id.trim() === "") {
		errors.push({ field: "opening_id", message: "Must be a non-empty string" });
	}
	if (
		typeof r.agency_org_domain !== "string" ||
		r.agency_org_domain.trim() === ""
	) {
		errors.push({
			field: "agency_org_domain",
			message: "Must be a non-empty string",
		});
	}
	return errors;
}

export function validateRemoveOpeningAgencyRequest(
	req: unknown
): ValidationError[] {
	return validateAssignOpeningAgencyRequest(req);
}

export function validateListOpeningAgenciesRequest(
	req: unknown
): ValidationError[] {
	const r = reqObj(req);
	if (!r) return [{ field: "$root", message: "Request body is required" }];
	const errors: ValidationError[] = [];
	if (typeof r.opening_id !== "string" || r.opening_id.trim() === "") {
		errors.push({ field: "opening_id", message: "Must be a non-empty string" });
	}
	return errors;
}

function validateOptionalPagination(
	r: Record<string, unknown>
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (r.limit !== undefined) {
		if (typeof r.limit !== "number" || r.limit < 1 || r.limit > 100) {
			errors.push({ field: "limit", message: "Must be between 1 and 100" });
		}
	}
	if (r.pagination_key !== undefined && typeof r.pagination_key !== "string") {
		errors.push({ field: "pagination_key", message: "Must be a string" });
	}
	return errors;
}

export function validateListAssignedOpeningsRequest(
	req: unknown
): ValidationError[] {
	const r = reqObj(req);
	if (!r) return [{ field: "$root", message: "Request body is required" }];
	return validateOptionalPagination(r);
}

export function validateListAgencyReferralsRequest(
	req: unknown
): ValidationError[] {
	const r = reqObj(req);
	if (!r) return [{ field: "$root", message: "Request body is required" }];
	return validateOptionalPagination(r);
}

export function validateReferCandidateRequest(req: unknown): ValidationError[] {
	const r = reqObj(req);
	if (!r) return [{ field: "$root", message: "Request body is required" }];
	const errors: ValidationError[] = [];
	if (typeof r.opening_id !== "string" || r.opening_id.trim() === "") {
		errors.push({ field: "opening_id", message: "Must be a non-empty string" });
	}
	if (
		typeof r.candidate_handle !== "string" ||
		r.candidate_handle.trim() === ""
	) {
		errors.push({
			field: "candidate_handle",
			message: "Must be a non-empty string",
		});
	}
	if (r.statement_text !== undefined) {
		if (typeof r.statement_text !== "string") {
			errors.push({ field: "statement_text", message: "Must be a string" });
		} else if (r.statement_text.length > 2000) {
			errors.push({
				field: "statement_text",
				message: "Must be at most 2000 characters",
			});
		}
	}
	return errors;
}

File: specs/typespec/org/agency-referrals.go
package org

import (
	"vetchium-api-server.typespec/common"
)

// ApplicationMode controls who may apply to an opening.
type ApplicationMode string

const (
	ApplicationModeOpen       ApplicationMode = "open"
	ApplicationModeAgencyOnly ApplicationMode = "agency_only"
)

// AgencyReferralState is the lifecycle state of an agency referral.
type AgencyReferralState string

const (
	AgencyReferralStatePending         AgencyReferralState = "pending"
	AgencyReferralStateAcceptedApplied AgencyReferralState = "accepted_applied"
	AgencyReferralStateDeclined        AgencyReferralState = "declined"
	AgencyReferralStateExpired         AgencyReferralState = "expired"
	AgencyReferralStateNotSelected     AgencyReferralState = "not_selected"
)

// ---- Consumer side: assign / list / remove agencies on an opening ----

type AssignOpeningAgencyRequest struct {
	OpeningID       string `json:"opening_id"`
	AgencyOrgDomain string `json:"agency_org_domain"`
}

type RemoveOpeningAgencyRequest struct {
	OpeningID       string `json:"opening_id"`
	AgencyOrgDomain string `json:"agency_org_domain"`
}

type ListOpeningAgenciesRequest struct {
	OpeningID string `json:"opening_id"`
}

type OpeningAgency struct {
	AgencyOrgDomain string `json:"agency_org_domain"`
	AgencyOrgName   string `json:"agency_org_name"`
	AssignedAt      string `json:"assigned_at"`
	ReferralsMade   int32  `json:"referrals_made"`
}

type ListOpeningAgenciesResponse struct {
	Agencies []OpeningAgency `json:"agencies"`
}

// ---- Agency side: openings I'm assigned to ----

type ListAssignedOpeningsRequest struct {
	PaginationKey *string `json:"pagination_key,omitempty"`
	Limit         *int32  `json:"limit,omitempty"`
}

type AssignedOpening struct {
	OpeningID         string `json:"opening_id"`
	ConsumerOrgDomain string `json:"consumer_org_domain"`
	OpeningNumber     int32  `json:"opening_number"`
	Title             string `json:"title"`
	AssignedAt        string `json:"assigned_at"`
}

type ListAssignedOpeningsResponse struct {
	Openings          []AssignedOpening `json:"openings"`
	NextPaginationKey *string           `json:"next_pagination_key,omitempty"`
}

// ---- Agency side: refer a candidate ----

type ReferCandidateRequest struct {
	OpeningID       string  `json:"opening_id"`
	CandidateHandle string  `json:"candidate_handle"`
	StatementText   *string `json:"statement_text,omitempty"`
}

type ReferCandidateResponse struct {
	ReferralID string `json:"referral_id"`
}

// ---- Agency side: referrals my agency has made ----

type ListAgencyReferralsRequest struct {
	PaginationKey *string `json:"pagination_key,omitempty"`
	Limit         *int32  `json:"limit,omitempty"`
}

type AgencyReferral struct {
	ReferralID        string              `json:"referral_id"`
	CandidateHandle   string              `json:"candidate_handle"`
	ConsumerOrgDomain string              `json:"consumer_org_domain"`
	OpeningNumber     int32               `json:"opening_number"`
	OpeningTitle      string              `json:"opening_title"`
	State             AgencyReferralState `json:"state"`
	CreatedAt         string              `json:"created_at"`
}

type ListAgencyReferralsResponse struct {
	Referrals         []AgencyReferral `json:"referrals"`
	NextPaginationKey *string          `json:"next_pagination_key,omitempty"`
}

func (r AssignOpeningAgencyRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.OpeningID == "" {
		errs = append(errs, common.ValidationError{Field: "opening_id", Message: "Must be a non-empty string"})
	}
	if r.AgencyOrgDomain == "" {
		errs = append(errs, common.ValidationError{Field: "agency_org_domain", Message: "Must be a non-empty string"})
	}
	return errs
}

func (r RemoveOpeningAgencyRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.OpeningID == "" {
		errs = append(errs, common.ValidationError{Field: "opening_id", Message: "Must be a non-empty string"})
	}
	if r.AgencyOrgDomain == "" {
		errs = append(errs, common.ValidationError{Field: "agency_org_domain", Message: "Must be a non-empty string"})
	}
	return errs
}

func (r ListOpeningAgenciesRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.OpeningID == "" {
		errs = append(errs, common.ValidationError{Field: "opening_id", Message: "Must be a non-empty string"})
	}
	return errs
}

func (r ListAssignedOpeningsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.Limit != nil && (*r.Limit < 1 || *r.Limit > 100) {
		errs = append(errs, common.ValidationError{Field: "limit", Message: "Must be between 1 and 100"})
	}
	return errs
}

func (r ReferCandidateRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.OpeningID == "" {
		errs = append(errs, common.ValidationError{Field: "opening_id", Message: "Must be a non-empty string"})
	}
	if r.CandidateHandle == "" {
		errs = append(errs, common.ValidationError{Field: "candidate_handle", Message: "Must be a non-empty string"})
	}
	if r.StatementText != nil && len(*r.StatementText) > 2000 {
		errs = append(errs, common.ValidationError{Field: "statement_text", Message: "Must be at most 2000 characters"})
	}
	return errs
}

func (r ListAgencyReferralsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.Limit != nil && (*r.Limit < 1 || *r.Limit > 100) {
		errs = append(errs, common.ValidationError{Field: "limit", Message: "Must be between 1 and 100"})
	}
	return errs
}
